src/freenet/io/Inet6AddressMatcher.java

```diff
@@ -66,18 +64,67 @@ public Inet6AddressMatcher(String pattern) {
 		}
 	}
 
-	private byte[] convertToBytes(String address) {
-		StringTokenizer addressTokens = new StringTokenizer(address, ":");
-		if (addressTokens.countTokens() != 8) {
-			throw new IllegalArgumentException(address + " is not an IPv6 address.");
-		}
+	private byte[] convertToBytes(String address) throws IllegalArgumentException {
+		String[] addressTokens = address.split(":");
+		int tokenPosition = 0;
 		byte[] addressBytes = new byte[16];
+		byte[] addressBytesEnd = new byte[16];
 		int count = 0;
-		while (addressTokens.hasMoreTokens()) {
-			int addressWord = Integer.parseInt(addressTokens.nextToken(), 16);
-			addressBytes[count * 2] = (byte) ((addressWord >> 8) & 0xff);
-			addressBytes[count * 2 + 1] = (byte) (addressWord & 0xff);
-			count++;
+		int endCount = -1;
+		if(address.startsWith(":")) {
+			if(address.startsWith("::")) {
+				if(address == "::") {
+					return addressBytes;
+				}
+				tokenPosition+=2; // This is an empty token, could be mistaken as duplicate ::
+				endCount = 0;
+			} else {
+				throw new IllegalArgumentException(address + " is not an IPv6 address.");
+			}
+		}
+		while (tokenPosition < addressTokens.length) {
+			String token = addressTokens[tokenPosition];
+			tokenPosition++;
+			if(!token.isEmpty() ) {
+				int addressWord;
+				try {
+					addressWord = Integer.parseInt(token, 16);
+				} catch(NumberFormatException e) {
+					throw new IllegalArgumentException(address + " is not an IPv6 address.");
+				}
+				if(endCount == -1) {
+					if(count >= 8) {
+						throw new IllegalArgumentException(address + " is not an IPv6 address.");
+					}
+					addressBytes[count * 2] = (byte) ((addressWord >> 8) & 0xff);
+					addressBytes[count * 2 + 1] = (byte) (addressWord & 0xff);
+					count++;
+				} else {
+					if(count + endCount >= 7) {
+						throw new IllegalArgumentException(address + " is not an IPv6 address.");
+					}
+					addressBytesEnd[endCount * 2] = (byte) ((addressWord >> 8) & 0xff);
+					addressBytesEnd[endCount * 2 + 1] = (byte) (addressWord & 0xff);
+					endCount++;
+				}
+			} else if(endCount == -1) {
+				endCount = 0;
+			} else {
+				throw new IllegalArgumentException(address + " is not an IPv6 address.");
+			}
+		}
+		if(endCount != -1 && count + endCount >= 8) { // Maybe not needed
```

Comment:
What do you mean, “maybe?” 😄 

You have the tests (that look rather exhaustive), when running them with coverage you should see immediately if this code is necessary.

Reply:
I'm sure it is not needed, but there's no harm staying there, just in case. I have removed this part of code as you suggests.
The tests cannot enumerate all possible invalid inputs, and indeed some were missed, for example, IPV6 address with percentage scope and 1:2:3:4:5:6:7: is broken with AddressIdentifier and Inet6AddressMatcher respectively. I have fixed those bugs and added them into the test suite as well.